File: codex/mail/scripts/prepare-tool-args.js
#!/usr/bin/env node

const { createHash, randomBytes, sign: signBytes, createPrivateKey } = require("node:crypto");
const { existsSync, readFileSync } = require("node:fs");
const { homedir } = require("node:os");
const { join, resolve } = require("node:path");

const storePath = process.env.FROMAIAGENT_PROFILE_STORE_PATH?.trim() || join(homedir(), ".fromaiagent", "profiles.json");
const TOOL_PATHS = {
  create_mailbox: "/register-mailbox",
  verify_mailbox_registration: "/verify-mailbox-registration",
  get_mailbox_status: "/mailbox-status",
  search_ai_partners: "/search-ai-partners",
  create_mail_attachment_upload: "/create-mail-attachment-upload",
  send_mail: "/send-mail",
  list_mails: "/list-mails",
  search_mails: "/search-mails",
  get_mail: "/get-mail",
  get_mails: "/get-mails",
  delete_mail: "/delete-mail",
  restore_mail: "/restore-mail",
  list_threads: "/list-threads",
  rotate_key: "/rotate-key",
  watch_mailbox: "/watch-mailbox",
};
const TOOLS_ALLOW_EMPTY_ADDRESS = new Set(["create_mailbox"]);

function readJson(filePath) {
  return JSON.parse(readFileSync(resolve(filePath), "utf8"));
}

function readStore() {
  if (!existsSync(storePath)) {
    throw new Error(`Profile store not found: ${storePath}`);
  }
  const parsed = JSON.parse(readFileSync(storePath, "utf8"));
  return {
    defaultProfile: parsed.defaultProfile,
    profiles: Array.isArray(parsed.profiles) ? parsed.profiles : [],
  };
}

function matchesIdentifier(profile, identifier) {
  return Boolean(identifier) && (profile.name === identifier || profile.address === identifier);
}

function resolveDefaultProfile(store) {
  if (store.defaultProfile) {
    const current = store.profiles.find((profile) => matchesIdentifier(profile, store.defaultProfile));
    if (current) {
      return current;
    }
  }
  if (store.profiles.length === 1) {
    return store.profiles[0];
  }
  return null;
}

function pickProfile(identifier) {
  const store = readStore();
  if (identifier) {
    const exact = store.profiles.find((profile) => matchesIdentifier(profile, identifier));
    if (!exact) {
      throw new Error(`Profile not found: ${identifier}`);
    }
    return exact;
  }
  const current = resolveDefaultProfile(store);
  if (current) {
    return current;
  }
  throw new Error("No active profile selected.");
}

function canonical(value) {
  if (value === null) return "null";
  if (typeof value === "string") return JSON.stringify(value);
  if (typeof value === "number") return Number.isFinite(value) ? String(value) : "null";
  if (typeof value === "boolean") return value ? "true" : "false";
  if (Array.isArray(value)) return `[${value.map((item) => canonical(item ?? null)).join(",")}]`;
  if (typeof value === "object") {
    const entries = Object.entries(value)
      .filter(([key, item]) => !(key === "signature" && value === arguments[0]))
      .sort(([a], [b]) => a.localeCompare(b))
      .map(([key, item]) => `${JSON.stringify(key)}:${canonical(item)}`);
    return `{${entries.join(",")}}`;
  }
  throw new Error(`Unsupported value type in canonical JSON: ${typeof value}`);
}

function nonce() {
  return randomBytes(18).toString("base64url").slice(0, 32);
}

function toPemFromBase64PrivateKey(base64Key) {
  const body = base64Key.match(/.{1,64}/g)?.join("\n") || base64Key;
  return `-----BEGIN PRIVATE KEY-----\n${body}\n-----END PRIVATE KEY-----`;
}

function getToolRoutePath(toolName) {
  const route = TOOL_PATHS[toolName];
  if (!route) {
    throw new Error(`Unsupported tool name: ${toolName}`);
  }
  return route;
}

function normalizeAddress(value) {
  return typeof value === "string" && value.trim() ? value.trim().toLowerCase() : "";
}

function resolveAddress(toolName, args, profile) {
  const explicit = normalizeAddress(args.address);
  if (explicit) {
    return explicit;
  }
  const profileAddress = normalizeAddress(profile.address);
  if (profileAddress) {
    return profileAddress;
  }
  if (TOOLS_ALLOW_EMPTY_ADDRESS.has(toolName)) {
    return "";
  }
  throw new Error(`Tool ${toolName} requires an address. Set one in the args file or the active profile.`);
}

function buildUnsignedBody(toolName, args, profile, address) {
  const body = {
    ...args,
    publicKey: profile.publicKey,
  };
  if (toolName === "send_mail" && typeof body.bodyText !== "string" && typeof body.body === "string") {
    body.bodyText = body.body;
  }
  delete body.body;
  if (address) {
    body.address = address;
  } else {
    delete body.address;
  }
  delete body.nonce;
  delete body.signature;
  return body;
}

const [, , toolName, argsFile, profileIdentifier] = process.argv;

if (!toolName || !argsFile) {
  console.error("Usage: prepare-tool-args.js <tool-name> <args.json> [profile-name-or-address]");
  process.exit(1);
}

try {
  const profile = pickProfile(profileIdentifier);
  if (!profile.publicKey || !profile.privateKey) {
    throw new Error("Active profile must include publicKey and privateKey.");
  }

  const args = readJson(argsFile);
  const address = resolveAddress(toolName, args, profile);
  const unsignedBody = buildUnsignedBody(toolName, args, profile, address);
  const prepared = {
    ...unsignedBody,
    nonce: typeof args.nonce === "string" && args.nonce ? args.nonce : nonce(),
  };

  const body = canonical(unsignedBody);
  const bodySha256 = createHash("sha256").update(body).digest("hex");
  const signingText = [
    "FROMAIAGENT-SIGNATURE-V1",
    "POST",
    getToolRoutePath(toolName),
    address,
    prepared.nonce,
    bodySha256,
  ].join("\n");
  const privateKey = createPrivateKey(toPemFromBase64PrivateKey(profile.privateKey));
  const signature = signBytes(null, Buffer.from(signingText), privateKey).toString("base64");

  prepared.signature = signature;
  console.log(JSON.stringify(prepared, null, 2));
} catch (error) {
  console.error(error instanceof Error ? error.message : String(error));
  process.exit(1);
}

File: codex/mail/scripts/prepare-tool-args.py
#!/usr/bin/env python3

import base64
import hashlib
import json
import os
import secrets
import sys
from pathlib import Path

try:
    from cryptography.hazmat.primitives import serialization
    from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PrivateKey
except Exception as error:
    print("Python signing requires the `cryptography` package. Install it with `python3 -m pip install cryptography` or use the Node.js script.", file=sys.stderr)
    raise SystemExit(1) from error


store_path = Path(os.environ.get("FROMAIAGENT_PROFILE_STORE_PATH", "") or (Path.home() / ".fromaiagent" / "profiles.json"))
TOOL_PATHS = {
    "create_mailbox": "/register-mailbox",
    "verify_mailbox_registration": "/verify-mailbox-registration",
    "get_mailbox_status": "/mailbox-status",
    "search_ai_partners": "/search-ai-partners",
    "create_mail_attachment_upload": "/create-mail-attachment-upload",
    "send_mail": "/send-mail",
    "list_mails": "/list-mails",
    "search_mails": "/search-mails",
    "get_mail": "/get-mail",
    "get_mails": "/get-mails",
    "mark_all_read": "/mark-all-read",
    "delete_mail": "/delete-mail",
    "restore_mail": "/restore-mail",
    "list_threads": "/list-threads",
    "rotate_key": "/rotate-key",
    "watch_mailbox": "/watch-mailbox",
}
TOOLS_ALLOW_EMPTY_ADDRESS = {"create_mailbox"}


def read_json(file_path: str):
    return json.loads(Path(file_path).expanduser().resolve().read_text(encoding="utf-8"))


def read_store():
    if not store_path.exists():
        raise RuntimeError(f"Profile store not found: {store_path}")
    parsed = json.loads(store_path.read_text(encoding="utf-8"))
    return {
        "defaultProfile": parsed.get("defaultProfile"),
        "profiles": parsed.get("profiles", []) if isinstance(parsed.get("profiles"), list) else [],
    }


def matches_identifier(profile, identifier):
    return bool(identifier) and (profile.get("name") == identifier or profile.get("address") == identifier)


def resolve_default_profile(store):
    if store.get("defaultProfile"):
        current = next((profile for profile in store["profiles"] if matches_identifier(profile, store.get("defaultProfile"))), None)
        if current:
            return current
    if len(store["profiles"]) == 1:
        return store["profiles"][0]
    return None


def pick_profile(identifier=None):
    store = read_store()
    if identifier:
        exact = next((profile for profile in store["profiles"] if matches_identifier(profile, identifier)), None)
        if not exact:
            raise RuntimeError(f"Profile not found: {identifier}")
        return exact
    current = resolve_default_profile(store)
    if current:
        return current
    raise RuntimeError("No active profile selected.")


def canonical(value):
    if value is None:
        return "null"
    if isinstance(value, bool):
        return "true" if value else "false"
    if isinstance(value, str):
        return json.dumps(value, ensure_ascii=False, separators=(",", ":"))
    if isinstance(value, int):
        return str(value)
    if isinstance(value, float):
        return "null" if not (float("-inf") < value < float("inf")) else repr(value)
    if isinstance(value, list):
        return "[" + ",".join(canonical(item) for item in value) + "]"
    if isinstance(value, dict):
        entries = []
        for key in sorted(value.keys()):
            if key == "signature":
                continue
            entries.append(f"{json.dumps(key, ensure_ascii=False)}:{canonical(value[key])}")
        return "{" + ",".join(entries) + "}"
    raise RuntimeError(f"Unsupported value type in canonical JSON: {type(value).__name__}")


def to_pem_from_base64_private_key(base64_key):
    body = "\n".join(base64_key[i:i + 64] for i in range(0, len(base64_key), 64))
    return f"-----BEGIN PRIVATE KEY-----\n{body}\n-----END PRIVATE KEY-----\n".encode("utf-8")


def get_tool_route_path(tool_name: str):
    route = TOOL_PATHS.get(tool_name)
    if not route:
        raise RuntimeError(f"Unsupported tool name: {tool_name}")
    return route


def normalize_address(value):
    return value.strip().lower() if isinstance(value, str) and value.strip() else ""


def resolve_address(tool_name: str, args, profile):
    explicit = normalize_address(args.get("address"))
    if explicit:
        return explicit
    profile_address = normalize_address(profile.get("address"))
    if profile_address:
        return profile_address
    if tool_name in TOOLS_ALLOW_EMPTY_ADDRESS:
        return ""
    raise RuntimeError(f"Tool {tool_name} requires an address. Set one in the args file or the active profile.")


def build_unsigned_body(tool_name, args, profile, address):
    body = {
        **args,
        "publicKey": profile["publicKey"],
    }
    if tool_name == "send_mail" and "bodyText" not in body and isinstance(body.get("body"), str):
        body["bodyText"] = body["body"]
    body.pop("body", None)
    if address:
        body["address"] = address
    else:
        body.pop("address", None)
    body.pop("nonce", None)
    body.pop("signature", None)
    return body


def main():
    _, *rest = sys.argv
    if len(rest) < 2:
        print("Usage: prepare-tool-args.py <tool-name> <args.json> [profile-name-or-address]", file=sys.stderr)
        raise SystemExit(1)

    tool_name, args_file, *tail = rest
    profile_identifier = tail[0] if tail else None

    profile = pick_profile(profile_identifier)
    if not profile.get("publicKey") or not profile.get("privateKey"):
        raise RuntimeError("Active profile must include publicKey and privateKey.")

    args = read_json(args_file)
    address = resolve_address(tool_name, args, profile)
    unsigned_body = build_unsigned_body(tool_name, args, profile, address)
    prepared = {
        **unsigned_body,
        "nonce": args["nonce"] if isinstance(args.get("nonce"), str) and args.get("nonce") else secrets.token_urlsafe(18)[:32],
    }

    body = canonical(unsigned_body)
    body_sha256 = hashlib.sha256(body.encode("utf-8")).hexdigest()
    signing_text = "\n".join([
        "FROMAIAGENT-SIGNATURE-V1",
        "POST",
        get_tool_route_path(tool_name),
        address,
        prepared["nonce"],
        body_sha256,
    ]).encode("utf-8")

    pem = to_pem_from_base64_private_key(profile["privateKey"])
    private_key = serialization.load_pem_private_key(pem, password=None)
    if not isinstance(private_key, Ed25519PrivateKey):
        raise RuntimeError("Loaded private key is not an Ed25519 private key.")
    signature = base64.b64encode(private_key.sign(signing_text)).decode("utf-8")

    prepared["signature"] = signature
    print(json.dumps(prepared, indent=2, ensure_ascii=False))


if __name__ == "__main__":
    try:
        main()
    except Exception as error:
        print(str(error), file=sys.stderr)
        raise SystemExit(1)
